app/questionnaire/routing/operations.py

```diff
@@ -81,18 +98,39 @@ def evaluate_any_in(lhs: Sequence, rhs: Sequence) -> bool:
 
     @staticmethod
     def resolve_date_from_string(
-        date_string: Optional[str], offset: Optional[dict[str, int]] = None
+        date_string: Optional[str],
+        offset: Optional[DateOffset] = None,
+        offset_by_full_weeks: bool = False,
     ) -> Optional[date]:
-        datetime_value = (
-            datetime.now(timezone.utc)
-            if date_string == "now"
-            else convert_to_datetime(date_string)
-        )
-        value_as_date = datetime_value.date() if datetime_value else None
-
-        if offset and value_as_date:
+        datetime_value = parse_datetime(date_string)
+        if not datetime_value:
+            return None
+
+        value_as_date = datetime_value.date()
+
+        if offset:
+            days_offset = offset.get("days", 0)
+
+            if day_of_week_offset := offset.get("day_of_week", None):
+                if 0 > days_offset > -7:
+                    raise ValueError(
+                        "Negative days offset must be less than or equal to -7 when used with `day_of_week` offset"
+                    )
+
+                days_difference = (
+                    value_as_date.weekday() - DAYS_OF_WEEK[day_of_week_offset]
+                )
+                days_to_reduce = days_difference % 7
+
+                if not offset_by_full_weeks and days_offset < 0 and days_difference < 0:
```

Comment:
This sentence is too complicate for me, we also shouldn't be adjusting the date, we should be changing the days_offset and then taking that off the date once it is right.  
```
                if not offset_by_full_weeks and days_offset < 0 and days_difference < 0:
                    # If we have a negative offset and date difference is less than 6 days,
                    # then the days we reduce the offset must be adjusted by 7
                    days_to_reduce -= 7
                days_offset -= days_to_reduce
```
I still haven't got the comment right

Reply:
Ya, had it like this initially but changed it to avoid the double negative. Don't mind changing it back if it is not confusing. 

As for the comment, ya, it is a tricky one to get your head around. Maybe an example would help if it was inline with the comments?. Ill come back to the comments once more have looked at it.

Reply:
I've redone the comment. Let me know if it is clearer, and what else we may want to add to it.

Reply:
I actually understand what it is doing, but it is difficult to write a comment, I failed, so can't complain whatever you write

Reply:
Double negatives are fine, coming from someone with a Maths degree

Reply:
Ugh I can't comment on this line and make it visible unless I unresolve the conversation - but it's nothing to do with the convo.

Could you just help out a future person reading this by parenthesising the terms - I'm pretty sure logic is left associative but I always find myself having to double check when maintaining, especially when mixing negatives and ands/ors

Reply:
Do you mean line 125 @jamesnunn?. I didn't add parentheses, since it would short circuit + it is ordered based on precedence. Happy to change if you think it will help:
```py
                if not offset_by_full_weeks and (
                    days_offset < 0 and days_difference < 0
                ):
```
Happy with this? ☝️

Reply:
Yeah that's what I'm after. How does it short circuit though?

Reply:
`and` in python short circuits, so if `not offset_by_full_weeks` is `True`, `days_offset < 0 and days_difference < 0` wouldn't even be evaluated. Is that what you were asking or?

Reply:
I'm not sure that's a problem because they're all `and`s and you'd be testing the result of that - it either gets to that block or it doesn't - all of the conditions are eventually evaluated. Anyway, it's fine in my opinion as it is!

Reply:
Not sure I follow. `all of the conditions are eventually evaluated`, only if leading conditions are True?. Anyhow, I decided to commit it as I don't mind, and it doesn't make it any less clearer. Plus, I already committed it for my other branch already 😄

Reply:
Yeah no worries we're splitting hairs again on this one. if you're happy with it then I am too